core/src/main/java/com/datastrato/gravitino/storage/relational/service/TableMetaService.java

```diff
@@ -0,0 +1,190 @@
+
+    TablePO tablePO = getTablePOBySchemaIdAndName(schemaId, identifier.name());
+
+    return POConverters.fromTablePO(tablePO, identifier.namespace());
+  }
+
+  public List<TableEntity> listTablesByNamespace(Namespace namespace) {
+    Namespace.checkTable(namespace);
+
+    Long schemaId = CommonMetaService.getInstance().getParentEntityIdByNamespace(namespace);
+
+    List<TablePO> tablePOs =
+        SessionUtils.getWithoutCommit(
+            TableMetaMapper.class, mapper -> mapper.listTablePOsBySchemaId(schemaId));
+
+    return POConverters.fromTablePOs(tablePOs, namespace);
+  }
+
+  public void insertTable(TableEntity tableEntity, boolean overwrite) {
+    try {
+      NameIdentifier.checkTable(tableEntity.nameIdentifier());
+
+      TablePO.Builder builder = new TablePO.Builder();
+      fillTablePOBuilderParentEntityId(builder, tableEntity.namespace());
+
+      SessionUtils.doWithCommit(
+          TableMetaMapper.class,
+          mapper -> {
+            TablePO po = POConverters.initializeTablePOWithVersion(tableEntity, builder);
+            if (overwrite) {
+              mapper.insertTableMetaOnDuplicateKeyUpdate(po);
+            } else {
+              mapper.insertTableMeta(po);
+            }
+          });
+    } catch (RuntimeException re) {
+      ExceptionUtils.checkSQLConstraintException(
+          re, Entity.EntityType.TABLE, tableEntity.nameIdentifier().toString());
+      throw re;
+    }
+  }
+
+  public <E extends Entity & HasIdentifier> TableEntity updateTable(
+      NameIdentifier identifier, Function<E, E> updater) throws IOException {
+    NameIdentifier.checkTable(identifier);
+
+    String tableName = identifier.name();
+
+    Long schemaId =
+        CommonMetaService.getInstance().getParentEntityIdByNamespace(identifier.namespace());
+
+    TablePO oldTablePO = getTablePOBySchemaIdAndName(schemaId, tableName);
+    TableEntity oldTableEntity = POConverters.fromTablePO(oldTablePO, identifier.namespace());
+    TableEntity newEntity = (TableEntity) updater.apply((E) oldTableEntity);
+    Preconditions.checkArgument(
+        Objects.equals(oldTableEntity.id(), newEntity.id()),
+        "The updated table entity id: %s should be same with the table entity id before: %s",
+        newEntity.id(),
+        oldTableEntity.id());
+
+    Integer updateResult;
+    try {
+      updateResult =
+          SessionUtils.doWithCommitAndFetchResult(
+              TableMetaMapper.class,
+              mapper ->
+                  mapper.updateTableMeta(
+                      POConverters.updateTablePOWithVersion(oldTablePO, newEntity), oldTablePO));
+    } catch (RuntimeException re) {
+      ExceptionUtils.checkSQLConstraintException(
+          re, Entity.EntityType.TABLE, newEntity.nameIdentifier().toString());
+      throw re;
+    }
+
+    if (updateResult > 0) {
+      return newEntity;
+    } else {
+      throw new IOException("Failed to update the entity: " + identifier);
+    }
+  }
```

Comment:
This update operation seems not transactional, but it's fine as we have locks in the outside.

Reply:
I learned that the current plan is to implement locks in the upper layer, because backends without transaction capabilities may be supported, so more complex locks are not implemented here.